### ISSUE TITLE
S4457 conflicting with async best practices

### ISSUE THREAD

**Issue by MxNbrt** [SonarSource/sonar-dotnet#6132]
S4457 tells me that my async code should be splitted into synchronous data validation and asynchronous data processing. E.g the compliant solution looks like this:

```
public static Task SkipLinesAsync(this TextReader reader, int linesToSkip)
{
    if (reader == null) { throw new ArgumentNullException(nameof(reader)); }
    if (linesToSkip < 0) { throw new ArgumentOutOfRangeException(nameof(linesToSkip)); }

    return reader.SkipLinesInternalAsync(linesToSkip);
}
```

My problem here is that last line which returns the Task directly instead of awaiting it. This conflicts with Microsofts recommendation 
 ["Prefer async/await over directly returning Task"](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-asyncawait-over-directly-returning-task) for async code which we followed everywhere in our code

So the correct solution according to Microsoft would be to make the whole method async and return await reader.SkipLinesInternalAsync(linesToSkip). After i got confused, I did what every developer would do and ended up at this [stackoverflow answer](https://stackoverflow.com/a/56909963/7291459) which also claimed that the rule is "overly cautious".

So now I am not quite sure which coding style to follow and whether the rule S4457 is even reasonable. 
Can you please clearify what is the purpose of S4457 and which code style would be best to follow?

Thank you very much

**Comment by rjgotten** [comment]
The rule _is_ overly cautious.
All code up to the first `await` is always executed synchronously, i.e. any `ArgumentException` or derivative that is thrown before the first `await` will execute synchronously. This isn't just an implementation detail. Several parts of and places in e.g. ASP.NET actually _rely_ on this behavior. (Notably the hosted services feature is such a place.)

If the criterium is fail-fast, then the correct rule implementation would be to not allow `ArgumentException` passed the first `await` - only before it.

**Comment by andrei-epure-sonarsource** [comment]
Thank you for opening this issue, @MxNbrt  . Your feedback helps us improve our products. And thank you, @rjgotten , for bringing the additional details.

We will look at it as soon as possible and come up with an answer.

**Comment by gregory-paidis-sonarsource** [comment]
Hey @MxNbrt ,
> Can you please clearify what is the purpose of [S4457](https://sonarsource.github.io/rspec/#/rspec/S4457/csharp) and which code style would be best to follow?

Here is an example:

```csharp
DoWorkAsync(null);          // fire-and-forget

// async-await used,
// treated as a special method for the async state machine,
// re-written by the compiler.
static async Task DoWorkAsync(int? maybeValue)
{
    if (!maybeValue.HasValue)
    {
        throw new ArgumentNullException(nameof(maybeValue));
    }

    await Task.Delay(42); // Actual work happens here
}
```

In this snippet, the "fire-and-forget" line will not throw an exception.

This is because exceptions raised in asynchronous methods are placed on the returned task.

If said task is not awaited, the exception is also not propagated to the call site, which could lead to potential bugs.

The correct approach would be something like the following:

```csharp
DoWorkAsync(null);          // fire-and-forget

// no async-await used, treated as a normal method
static Task DoWorkAsync(int? maybeValue)
{
    if (!maybeValue.HasValue)
    {
        throw new ArgumentNullException(nameof(maybeValue));
    }

    return Task.Delay(42); // Actual work happens here
}
```

In this case, the fire-and-forget method call will raise as expected.

> "Prefer async/await over directly returning Task" for async code which we followed everywhere in our code

Notice that the `await DoWorkAsync(null);` method call raises in both circumstances, so if your codebase is using "async/await all the way down" as you mentioned, as well as not exposing any public APIs that a third-party application might use (without using async/await), you can safely ignore this rule or mark it as "won't fix". 


I hope the snippets provided above clarify the utility of this rule.

**Comment by rjgotten** [comment]
The 'but fire-and-forget tasks' argument is, respectfully, bullshit.

The whole _point_ of fire-and-forget is that you kick something off, _and then forget about it_.
If you need to care about the arguments and need to care that the task actually completes succesfully, then you are using _the wrong pattern_.

Meanwhile, to pander to a very specific corner case that maybe makes up 1% of the total case, this style rule enforces work on 99% of the other cases.  To needlessly restructure code to comply with the rule's silly wishes; to explicitly mark each individual occurence of a rule violation as 'won't fix' in a central Sonarqube instance; or to add suppressions to each individual occurence in code. It's a rule that, when activivated, has in practicality a 99% false positive ratio needing workarounds.
And rules that have a 99% false positive ratio _aren't very good rules_.

Moreover, the proposed fix along `ArgumentException` and derivatives to be thrown synchronously, _relies on an internal implementation detail_ of the async/await state machine. The fact that code runs synchronously until the first `await` statement is an artifact of Microsoft's _current_ implementation. And that implementation is afaik _not_ guaranteed to remain. There have been extensive rewrites to the state machine generating code in the past; both for general performance reasons; as well as for added features such as when support for the lightweight `ValueTask` type was baked in.

In fact, it's been possible for _years_ for users _to write their own async method builder_ and to [tell the runtime to use that](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.asyncmethodbuilderattribute). Unity, among others, does this so it can tune performance characteristics to be more in line with what you'd expect from a game engine.

In other words: your proposed style improvement hingers on an internal implementation detail _you cannot guarantee_.

The **only** way you can guarantee receiving an exception back, _is to await the task_.

**Comment by MxNbrt** [comment]
Hello @gregory-paidis-sonarsource 

Thanks for clarifying, what your thoughts were when implementing the rule. Nonetheless i have to agree to @rjgotten and think that the fire-and-forget pattern should itself be a rule that is being violated here. The fire-and-forget pattern is bad and should only be used in a very small percentage of use cases. 

Thats why you should get rid of S4457 or at least disable it as a default. 

Thanks and have a nice day

**Comment by gregory-paidis-sonarsource** [comment]
Hello again @MxNbrt and @rjgotten,

After some testing, as well as internal discussion with the team, we decided to remove this rule from the `SonarWay` profile.
